### PR TITLE
Add capability to backfill missing days to upload.py

### PR DESCRIPTION
**Pull request by aiuto**
Sort of a major refactoring.

### CODE REVIEW

File: upload.py [hunk 1]
```diff
@@ -94,89 +97,152 @@ def GatherPreviousDownloads(connection, trailing_days):
...
+              filler.downloads))
+          self.add_daily_counts(filler)
+          downloads -= inc_downloads
+          sha256 -= inc_sha256
+          sig -= inc_sig
+
+      # assert: no need to backfill: downloads is what we got today
+      # assert: with backfill: downloads holds ~avg/day change over period
+      # assert: skip backfill: downloads holds multi-day delta
+      if (downloads > 10) and (previous.downloads * 115 // 100 < downloads):
```
Comment (jin) [inline upload.py]: extract magic numbers into constants

File: upload.py [hunk 2]
```diff
@@ -94,89 +97,152 @@ def GatherPreviousDownloads(connection, trailing_days):
...
+            downloads=0,
+            sha256=0,
+            sig=0)
+        self.process_sample(sample)
+
+    if not self.dry_run:
+      self.cursor.close()
+      self.connection.commit()
+
+  def process_sample(self, sample):
```
Comment (jin) [inline upload.py:146]: please write a comment on what this function is processing.

File: upload.py [hunk 3]
```diff
@@ -94,89 +97,152 @@ def GatherPreviousDownloads(connection, trailing_days):
...
+    sample_dt = datetime.datetime.strptime(sample.sample_date, '%Y-%m-%d')
+    sample_day = (sample_dt - _EPOCH).days
+    downloads = sha256 = sig= 0
```
Comment (jin) [inline upload.py]:
```suggestion
    downloads = sha256 = sig = 0
```

File: upload_test.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    delta_sha256 = 60
+    delta_sig = 20
+
+    s = upload.DownloadSample(
+        sample_date=sample_date,
+        file=file,
+        product=product,
+        version=version,
+        downloads_total=50 + delta_downloads * (n_backfill + 1) + 1,
+        sha256_total=30 + delta_sha256 * (n_backfill  + 1)+ 2,
```
Comment (jin) [inline upload_test.py]:
```suggestion
        sha256_total=30 + delta_sha256 * (n_backfill  + 1) + 2,
```

**Review by jin — APPROVED** [review]
Sorry for the long time on this review, had to backtrack the code to see what this was doing.

File: upload.py (hunk 1 above)
Comment (aiuto) [reply upload.py]: Done

File: upload.py (hunk 2 above)
Comment (aiuto) [reply upload.py:146]: done

**Pull request merged**